### PR TITLE
[bitnami/sonarqube] Close gap with upstream SonarQube Helm chart

### PR DESCRIPTION
**Pull request by aoterolorenzo**
Signed-off-by: Alberto Otero Lorenzo <lorenzoa@vmware.com>

### Description of the change

- [x] Adds a feature in order to be able to install plugins from the chart
- [x] Sets logic to include CACerts and parse them into SONARQUBE_WEB_JVM_OPTS and SONAR_CE_JAVAOPTS env vars

### Checklist

- [x] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/). This is *not necessary* when the changes only affect README.md files.
- [x] Variables are documented in the values.yaml and added to the `README.md` using [readme-generator-for-helm](https://github.com/bitnami-labs/readme-generator-for-helm)
- [x] Title of the pull request follows this pattern [bitnami/<name_of_the_chart>] Descriptive title
- [x] All commits signed off and in agreement of [Developer Certificate of Origin (DCO)](https://github.com/bitnami/charts/blob/main/CONTRIBUTING.md#sign-your-work)

### CODE REVIEW

**Comment by aoterolorenzo** [comment]
Plugins provisioning working as expected with below values:

```
plugins:
  install: [ "https://github.com/AmadeusITGroup/sonar-stash/releases/download/1.3.0/sonar-stash-plugin-1.3.0.jar" ]
```

Captura de Pantalla 2023-01-18 a las 18 49 45

Captura de Pantalla 2023-01-18 a las 18 50 04

**Comment by aoterolorenzo** [comment]
Ready to review! 

Example values for using the new features (I just used as an example and for easy testing an injection of the Spanish institution FNMT root certificate):

```
extraDeploy:
  - kind: Secret
    apiVersion: v1
    type: Opaque
    metadata:
      name: ca-certs-secret
      namespace: default
    data:
      cert.cer: |-
        MIIFgzCCA2ugAwIBAgIPXZONMGc2yAYdGsdUhGkHMA0GCSqGSIb3DQEBCwUAMDsxCzAJBgNVBAYT
        AkVTMREwDwYDVQQKDAhGTk1ULVJDTTEZMBcGA1UECwwQQUMgUkFJWiBGTk1ULVJDTTAeFw0wODEw
        MjkxNTU5NTZaFw0zMDAxMDEwMDAwMDBaMDsxCzAJBgNVBAYTAkVTMREwDwYDVQQKDAhGTk1ULVJD
        TTEZMBcGA1UECwwQQUMgUkFJWiBGTk1ULVJDTTCCAiIwDQYJKoZIhvcNAQEBBQADggIPADCCAgoC
        ggIBALpxgHpMhm5/yBNtwMZ9HACXjywMI7sQmkCpGreHiPibVmr75nuOi5KOpyVdWRHbNi63URcf
        qQgfBBckWKo3Shjf5TnUV/3XwSyRAZHiItQDwFj8d0fsjz50Q7qsNI1NOHZnjrDIbzAzWHFctPVr
        btQBULgTfmxKo0nRIBnuvMApGGWn3v7v3QqQIecaZ5JCEJhfTzC8PhxFtBDXaEAUwED653cXeuYL
        j2VbPNmaUtu1vZ5Gzz3rkQUCwJaydkxNEJY7kvqcfw+Z374jNUUeAlz+taibmSXaXvMiwzn15Cou
        08YfxGyqxRxqAQVKL9LFwag0Jl1mpdICIfkYtwb1TplvqKtMUejPUBjFd8g5CSxJkjKZqLsXF3mw
        WsXmo8RZZUc1g16p6DULmbvkzSDGm0oGObVo/CK67lWMK07q87Hj/LaZmtVC+nFNCM+HHmpxffnT
        tOmlcYF7wk5HlqX2doWjKI/pgG6BU6VtX7hI+cL5NqYuSf+4lsKMB7ObiFj86xsc3i1w4peSMKGJ
        47xVqCfWS+2QrYv6YyVZLag13cqXM7zlzced0ezvXg5KkAYmY6252TUtB7p2ZSysV4999AeU14EC
        ll2jB0nVetBX+RvnU0Z1qrB5QstocQjpYL05ac70r8NWQMetUqIJ5G+GR4of6ygnXYMgrwTJbFaa
        i0b1AgMBAAGjgYMwgYAwDwYDVR0TAQH/BAUwAwEB/zAOBgNVHQ8BAf8EBAMCAQYwHQYDVR0OBBYE
        FPd9xf3E6Jobd2Sn9R2gzL+HYJptMD4GA1UdIAQ3MDUwMwYEVR0gADArMCkGCCsGAQUFBwIBFh1o
        dHRwOi8vd3d3LmNlcnQuZm5tdC5lcy9kcGNzLzANBgkqhkiG9w0BAQsFAAOCAgEAB5BK3/MjTvDD
        nFFlm5wioooMhfNzKWtN/gHiqQxjAb8EZ6WdmF/9ARP67Jpi6Yb+tmLSbkyU+8B1RXxlDPiyN8+s
        D8+Nb/kZ94/sHvJwnvDKuO+3/3Y3dlv2bojzr2IyIpMNOmqOFGYMLVN0V2Ue1bLdI4E7pWYjJ2cJ
        j+F3qkPNZVEI7VFY/uY5+ctHhKQV8Xa7pO6kO8Rf77IzlhEYt8llvhjho6Tc+hj507wTmzl6NLrT
        Qfv6MooqtyuGC2mDOL7Nii4LcK2NJpLuHvUBKwrZ1pebbuCoGRw6IYsMHkCtA+fdZn71uSANA+iW
        +YJF1DngoABd15jmfZ5nc8OaKveri6E6FO80vFIOiZiaBECEHX5FaZNXzuvO+FB8TxxuBEOb+dY7
        Ixjp6o7RTUaN8Tvkasq6+yO3m/qZASlaWFot4/nUbQ4mrcFuNLwy+AwF+mWj2zs3gyLp1txyM/1d
        8iC9djwj2ij3+RvrWWTV3F9yfiD8zYm1kGdNYno/Tq0dwzn+evQoFt9B9kiABdcPUXmsEKvU7ANm
        5mqwujGSQkBqvjrTcuFqN1W8rB2Vt2lh8kORdOag0wokRqEIr9baRRmW1FMdW4R58MD3R++Lj8UG
        rp1MYp3/RgT408m2ECVAdf4WqslKYIYvuu8wd+RU4riEmViAqhOLUTpPSPaLtrM=
plugins:
  install: [ "https://github.com/AmadeusITGroup/sonar-stash/releases/download/1.6.0/sonar-stash-plugin-1.6.0.jar" ]
```

You can check if the certificate has been successful installed executing the following command on the sonarqube main pod CLI:

`echo changeit | keytool -list -v -keystore $(find $JAVA_HOME -name cacerts) | grep FNMT`

Ensuring the output is similar to:

```
Warning: use -cacerts option to access cacerts keystore
Enter keystore password:  Owner: OU=AC RAIZ FNMT-RCM, O=FNMT-RCM, C=ES
Issuer: OU=AC RAIZ FNMT-RCM, O=FNMT-RCM, C=ES
Owner: CN=AC RAIZ FNMT-RCM SERVIDORES SEGUROS, OID.2.5.4.97=VATES-Q2826004J, OU=Ceres, O=FNMT-RCM, C=ES
Issuer: CN=AC RAIZ FNMT-RCM SERVIDORES SEGUROS, OID.2.5.4.97=VATES-Q2826004J, OU=Ceres, O=FNMT-RCM, C=ES
```


Also, you can check the plugin provisioning through the CLI with the below command and output, or in the web interface accessible on port 80:

```
sonarqube@sonarqube-7467d6f7fb-b8xc9:/$ ls /opt/bitnami/sonarqube/extensions/plugins/
sonar-stash-plugin-1.6.0.jar
```

File: bitnami/sonarqube/templates/_helpers.tpl [hunk 1]
```diff
@@ -196,3 +210,52 @@ sonarqube: database
...
+*/}}
+{{- define "sonarqube.jvmCEOpts" -}}
+    {{- if .Values.caCerts.enabled -}}
+        {{ printf "-Djavax.net.ssl.trustStore=/bitnami/sonarqube/certs/cacerts %s" .Values.jvmCeOpts | trim | quote }}
+    {{- else -}}
+        {{ printf "" }}
+    {{- end -}}
+{{- end -}}
+
+
```
Comment (carrodher) [inline bitnami/sonarqube/templates/_helpers.tpl:261]:
```suggestion
```

File: bitnami/sonarqube/values.yaml [hunk 2]
```diff
@@ -102,6 +102,8 @@ sonarqubeUsername: user
 ## Defaults to a random 10-character alphanumeric string if not set
 ##
 sonarqubePassword: ""
+## @param provisioningFolder Directory to use for provisioning content to Sonarqube
+provisioningFolder: "/bitnami/sonarqube-provisioning"
```
Comment (carrodher) [inline bitnami/sonarqube/values.yaml:106]:
```suggestion
## @param provisioningFolder Directory to use for provisioning content to Sonarqube
##
provisioningFolder: "/bitnami/sonarqube-provisioning"
```

File: bitnami/sonarqube/values.yaml [hunk 3]
```diff
@@ -578,6 +586,112 @@ ingress:
   ##
   extraRules: []
 
+## @section SonarQube caCerts provisioning parameters
+##
+## Provide a secret containing one or more certificate files in the keys that will be added to cacerts
+## The cacerts file will be set via SONARQUBE_WEB_JVM_OPTS and SONAR_CE_JAVAOPTS
+caCerts:
+  ## @param caCerts.enabled Enable the use of caCerts
+  enabled: true
```
Comment (carrodher) [inline bitnami/sonarqube/values.yaml:595]:
```suggestion
  ## @param caCerts.enabled Enable the use of caCerts
  ##
  enabled: true
```

File: bitnami/sonarqube/values.yaml [hunk 4]
```diff
@@ -578,6 +586,112 @@ ingress:
...
+  ## @param caCerts.image.repository Bitnami Shell image repository
+  ## @param caCerts.image.tag Bitnami Shell image tag (immutable tags are recommended)
+  ## @param caCerts.image.digest Bitnami Shell image digest in the way sha256:aa.... Please note this parameter, if set, will override the tag
+  ## @param caCerts.image.pullPolicy Bitnami Shell image pull policy
+  ## @param caCerts.image.pullSecrets Bitnami Shell image pull secrets
+  ##
+  image:
+    registry: docker.io
+    repository: bitnami/bitnami-shell
+    tag: 11-debian-11-r63
```
Comment (carrodher) [inline bitnami/sonarqube/values.yaml:608]: Please note you need to define this container image in our internal test & release pipeline so the tag is automatically updated as the other images of this Helm chart

File: bitnami/sonarqube/values.yaml [hunk 5]
```diff
@@ -578,6 +586,112 @@ ingress:
...
+    ## Optionally specify an array of imagePullSecrets.
+    ## Secrets must be manually created in the namespace.
+    ## ref: https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/
+    ## e.g:
+    ## pullSecrets:
+    ##   - myRegistryKeySecretName
+    ##
+    pullSecrets: []
+  ## @param caCerts.secret Name of the secret containing the certificates
+  secret: ca-certs-secret
```
Comment (carrodher) [inline bitnami/sonarqube/values.yaml:620]:
```suggestion
  ## @param caCerts.secret Name of the secret containing the certificates
  ##
  secret: ca-certs-secret
```

File: bitnami/sonarqube/values.yaml [hunk 6]
```diff
@@ -578,6 +586,112 @@ ingress:
...
+  ## plugins:
+  ##  install:
+  ##    - "https://github.com/AmadeusITGroup/sonar-stash/releases/download/1.3.0/sonar-stash-plugin-1.3.0.jar"
+  ##    - "https://github.com/SonarSource/sonar-ldap/releases/download/2.2-RC3/sonar-ldap-plugin-2.2.0.601.jar"
+  ##
+  install: []
+  ## @param plugins.netrcCreds .netrc secret file with a key "netrc" to use basic auth while downloading plugins
+  netrcCreds: ""
+  ## @param plugins.noCheckCertificate Set to true to not validate the server's certificate to download plugin
+  noCheckCertificate: true
```
Comment (carrodher) [inline bitnami/sonarqube/values.yaml:653]:
```suggestion
  ## @param plugins.netrcCreds .netrc secret file with a key "netrc" to use basic auth while downloading plugins
  ##
  netrcCreds: ""
  ## @param plugins.noCheckCertificate Set to true to not validate the server's certificate to download plugin
  ##
  noCheckCertificate: true
```

File: bitnami/sonarqube/values.yaml [hunk 7]
```diff
@@ -578,6 +586,112 @@ ingress:
...
+  ## @param plugins.image.repository Bitnami Shell image repository
+  ## @param plugins.image.tag Bitnami Shell image tag (immutable tags are recommended)
+  ## @param plugins.image.digest Bitnami Shell image digest in the way sha256:aa.... Please note this parameter, if set, will override the tag
+  ## @param plugins.image.pullPolicy Bitnami Shell image pull policy
+  ## @param plugins.image.pullSecrets Bitnami Shell image pull secrets
+  ##
+  image:
+    registry: docker.io
+    repository: bitnami/bitnami-shell
+    tag: 11-debian-11-r63
```
Comment (carrodher) [inline bitnami/sonarqube/values.yaml:666]: Please note you need to define this container image in our internal test & release pipeline so the tag is automatically updated as the other images of this Helm chart

File: bitnami/sonarqube/templates/_helpers.tpl [hunk 8]
```diff
@@ -196,3 +210,50 @@ sonarqube: database
...
+    {{ printf "%s-install-plugins-initcontainer" (include "common.names.fullname" .) }}
+{{- end -}}
+
+{{/*
+Create the name of the deployment's caCerts initContainer
+*/}}
+{{- define "sonarqube.caCerts.initContainer" -}}
+    {{ printf "%s-ca-certs-initcontainer" (include "common.names.fullname" .) }}
+{{- end -}}
+
```
Comment (javsalgar) [inline bitnami/sonarqube/templates/_helpers.tpl:227]: I see that this is being used only once. Aren't we actually creating more lines than necessary by creating a macro?

File: bitnami/sonarqube/templates/_helpers.tpl [hunk 9]
```diff
@@ -196,3 +210,50 @@ sonarqube: database
...
+{{/*
+Create the name of the install_plugins configMap
+*/}}
+{{- define "sonarqube.installPlugins.configMap" -}}
```
Comment (javsalgar) [inline bitnami/sonarqube/templates/_helpers.tpl:231]: Same here

File: bitnami/sonarqube/templates/_helpers.tpl [hunk 10]
```diff
@@ -196,3 +210,50 @@ sonarqube: database
        externalDatabase.database=DB_NAME
 {{- end -}}
 {{- end -}}
+
+{{/*
+Create the name of the deployment's install_plugins initContainer
+*/}}
+{{- define "sonarqube.installPlugins.initContainer" -}}
+    {{ printf "%s-install-plugins-initcontainer" (include "common.names.fullname" .) }}
```
Comment (javsalgar) [inline bitnami/sonarqube/templates/_helpers.tpl:218]: Same here

**Review by javsalgar — CHANGES_REQUESTED** [review]

**Review by javsalgar — APPROVED** [review]

**Pull request merged**